File: js/jackknife/jackknife_recognizer.js
/**
 * Copyright 2017 the University of Central Florida Research Foundation, Inc.
 * All rights reserved.
 * 
 *     Eugene M. Taranta II <etaranta@gmail.com>
 *     Amirreza Samiei <samiei@knights.ucf.edu>
 *     Mehran Maghoumi <mehran@cs.ucf.edu>
 *     Pooya Khaloo <pooya@cs.ucf.edu>
 *     Corey R. Pittman <cpittman@knights.ucf.edu>
 *     Joseph J. LaViola Jr. <jjl@cs.ucf.edu>
 * 
 * Subject to the terms and conditions of the Florida Public Educational
 * Institution non-exclusive software license, this software is distributed 
 * under a non-exclusive, royalty-free, non-sublicensable, non-commercial, 
 * non-exclusive, academic research license, and is distributed without warranty
 * of any kind express or implied. 
 *
 * The Florida Public Educational Institution non-exclusive software license
 * is located at <https://github.com/ISUE/Jackknife/blob/master/LICENSE>.
 */

/**
 * Constructor.
 */
function Jackknife(blades) {
    /**
     * The set of measures and features used in this instance
     * of Jackknife.
     */
    this.blades = blades;

    /**
     * Array of all gesture templates.
     */
    this.templates = [];
}

/**
 * Turn sample into template.
 */
Jackknife.prototype.add_template = function(sample) {
    this.templates.push(new JackknifeTemplate(this.blades, sample));
}

/**
 * Determine gesture class of the given sample.
 * The gesture id is returned.
 *
 * This function can be modified to return an n-best list, but
 * because of early rejection, such a list may not make sense.
 *
 */
Jackknife.prototype.classify = function(trajectory) {

    // Handle non-sample data
    if (trajectory instanceof Sample) {
        return this.classify(trajectory.trajectory);
    }

    // extract important information from candidate sample
    var features = new JackknifeFeatures(this.blades, trajectory);

    // now find the lower bound for this candidate
    // compared against each template.
    // also cache correction factor scores
    var template_cnt = this.templates.length;

    for (var tt = 0; tt < template_cnt; tt++) {
        var cf = 1.0;

        if (this.blades.cf_abs_distance > 0) {
            cf *= 1.0 / Math.max(
                0.01,
                features.abs.dot(this.templates[tt].features.abs));
        }

        if (this.blades.cf_bb_widths > 0) {
            cf *= 1.0 / Math.max(
                0.01,
                features.bb.dot(this.templates[tt].features.bb));
        }

        this.templates[tt].cf = cf;

        if (this.blades.lower_bound > 0) {
            this.templates[tt].lb = cf * this.lower_bound(
                features.vecs,
                this.templates[tt]);
        }
    }

    // Now sort all templates based on their lower bound
    // results. This helps to improve culling later on.
    this.templates.sort(compareTemplates);

    // track best result so far
    var best = Number.POSITIVE_INFINITY;
    var ret = -1;

    for (var tt = 0; tt < template_cnt; tt++) {
        // cull based on lower bound vs rejection threshold
        if (this.templates[tt].lb > this.templates[tt].rejection_threshold)
            continue;

        // cull based on lower bound vs best score
        if (this.templates[tt].lb > best)
            continue;

        // forced to do full evaluation
        var score = this.templates[tt].cf;

        // note that the vecs can be either direction vectors
        // or points, depending on selected measure
        score *= this.DTW(
            features.vecs,
            this.templates[tt].features.vecs);
        // only accept if below rejection threshold
        if (score > this.templates[tt].rejection_threshold)
            continue;

        // save, if best result
        if (score < best) {
            best = score;
            ret = this.templates[tt].gesture_id;
        }
    }

    return ret;
}


/**
 * Learn a rejection threshold for each template.
 * Call after all templates have been added.
 *
 */
Jackknife.prototype.train = function(gpsr_n, gpsr_r, beta) {
    var template_cnt = this.templates.length;
    var distributions = [];
    var synthetic = [];

    var worst_score = 0.0;

    //
    // Create negative samples.
    //
    for (var ii = 0; ii < 1000; ii++) {
        synthetic.length = 0
        // Splice two samples together
        // to create one negative sample.
        for (var jj = 0; jj < 2; jj++) {
            var tt = Math.floor(Math.random() * template_cnt % template_cnt);

            var s = this.templates[tt].sample;
            var len = s.trajectory.length;

            var start = Math.floor(Math.random() * (len / 2) % (len / 2));

            for (var kk = 0; kk < len / 2; kk++) {
                synthetic.push(new Vector(s.trajectory[start + kk]));
            }
        }

        var features = new JackknifeFeatures(
            this.blades,
            synthetic);

        // and score it
        for (var tt = 0; tt < template_cnt; tt++) {
            var score = this.DTW(
                features.vecs,
                this.templates[tt].features.vecs);

            if (worst_score < score)
                worst_score = score;

            if (ii > 50)
                distributions[tt].add_negative_score(score);
        }

        // Generate a few samples to get an estimate
        // of worst possible score.
        if (ii != 50)
            continue;

        // allocate distributions
        for (var tt = 0; tt < template_cnt; tt++) {
            distributions.push(
                new Distributions(
                    worst_score,
                    1000));
        }
    }

    //
    // Create positive examples.
    //
    for (var tt = 0; tt < template_cnt; tt++) {
        for (var ii = 0; ii < 1000; ii++) {
            synthetic.length = 0;

            // Create a synthetic variation of the sample.
            gpsr(
                this.templates[tt].sample.trajectory,
                synthetic,
                gpsr_n,
                0.25,
                gpsr_r);

            var features = new JackknifeFeatures(this.blades, synthetic);
            // and score it
            var score = this.DTW(features.vecs, this.templates[tt].features.vecs);
            distributions[tt].add_positive_score(score);
        }
    }

    //
    // Now extract the rejection thresholds.
    //
    for (var tt = 0; tt < template_cnt; tt++) {
        var threshold = distributions[tt].rejection_threshold(beta);
        this.templates[tt].rejection_threshold = threshold;
    }
}

/**
 * The DTW algorithm. Awesome!
 */
Jackknife.prototype.DTW = function(v1, v2) {
    var cost = new Array(v1.length + 1);
    for (var i = 0; i < cost.length; i++)
        cost[i] = new Array(v2.length + 1);


    // initialize matrix
    for (var ii = 0; ii <= v1.length; ii++) {
        for (var jj = 0; jj <= v2.length; jj++) {
            cost[ii][jj] = Number.POSITIVE_INFINITY;

        }
    }

    cost[0][0] = 0.0;

    // using DP to find solution
    for (var ii = 1; ii <= v1.length; ii++) {
        for (var jj = Math.max(1, ii - Math.floor(this.blades.radius));
            jj <= Math.min(Math.floor(v2.length), ii + Math.floor(this.blades.radius));
            jj++) {
            // pick minimum cost path (neighbor) to
            // extend to this ii, jj element
            var minimum = Math.min(
                Math.min(
                    cost[ii - 1][jj], // repeat v1 element
                    cost[ii][jj - 1]), // repeat v2 element
                cost[ii - 1][jj - 1]); // don't repeat either

            cost[ii][jj] = minimum;

            if (this.blades.inner_product) {
                cost[ii][jj] += 1.0 - v1[ii - 1].dot(v2[jj - 1]);
            } else if (this.blades.euclidean_distance) {
                cost[ii][jj] += v1[ii - 1].l2norm2(v2[jj - 1]);
            } else {
                assert(0);
            }
        }
    }
    //console.log(cost[v1.length][v2.length])
    return cost[v1.length][v2.length];
}

/**
 * Find the lower bound score for a candidate
 * against a given template.
 */
Jackknife.prototype.lower_bound = function(vecs, t) {
    var lb = 0.0; // lower bound
    var component_cnt = vecs[0].data.length;

    for (var ii = 0; ii < vecs.length; ii++) {
        var cost = 0.0;

        for (var jj = 0; jj < component_cnt; jj++) {
            if (this.blades.inner_product) {
                if (vecs[ii].data[jj] < 0.0) {
                    cost += vecs[ii].data[jj] * t.lower[ii].data[jj];
                } else {
                    cost += vecs[ii].data[jj] * t.upper[ii].data[jj];
                }
            } else if (this.blades.euclidean_distance) {
                var diff = 0.0;

                if (vecs[ii].data[jj] < t.lower[ii].data[jj]) {
                    diff = vecs[ii].data[jj] - t.lower[ii].data[jj];
                } else if (vecs[ii].data[jj] > t.upper[ii].data[jj]) {
                    diff = vecs[ii].data[jj] - t.upper[ii].data[jj];
                }

                cost += (diff * diff);
            } else {
                assert(0);
            }
        }

        // inner products are bounded
        if (this.blades.inner_product) {
            cost = 1.0 - Math.min(1.0, Math.max(-1.0, cost));
        }

        lb += cost;
    }

    return lb;
}

File: csharp/Jackknife/JackknifeRecognizer.cs
/**
 * Copyright 2017 the University of Central Florida Research Foundation, Inc.
 * All rights reserved.
 * 
 *     Eugene M. Taranta II <etaranta@gmail.com>
 *     Amirreza Samiei <samiei@knights.ucf.edu>
 *     Mehran Maghoumi <mehran@cs.ucf.edu>
 *     Pooya Khaloo <pooya@cs.ucf.edu>
 *     Corey R. Pittman <cpittman@knights.ucf.edu>
 *     Joseph J. LaViola Jr. <jjl@cs.ucf.edu>
 * 
 * Subject to the terms and conditions of the Florida Public Educational
 * Institution non-exclusive software license, this software is distributed 
 * under a non-exclusive, royalty-free, non-sublicensable, non-commercial, 
 * non-exclusive, academic research license, and is distributed without warranty
 * of any kind express or implied. 
 *
 * The Florida Public Educational Institution non-exclusive software license
 * is located at <https://github.com/ISUE/Jackknife/blob/master/LICENSE>.
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Jackknife
{
    public class Jackknife
    {
        /**
         * The DTW cost matrix that is initialized once and reused multiple times.
         */
        private List<List<double>> cost = null;

        /**
         * Vector of all gesture templates.
         */
        private List<JackknifeTemplate> templates;

        /**
         * The set of measures and features used in this instance
         * of Jackknife.
         */
        public JackknifeBlades Blades { get; set; }

        /**
         * Constructor.
         */
        public Jackknife(JackknifeBlades blades)
        {
            if (!(blades.InnerProduct ^ blades.EuclideanDistance))
            {
                throw new ArgumentException("!(blades.InnerProduct ^ blades.EuclideanDistance)");
            }

            this.Blades = blades;
            templates = new List<JackknifeTemplate>();
        }

        /**
         * Turn sample into template.
         */
        public void AddTemplate(Sample sample)
        {
            JackknifeTemplate t = new JackknifeTemplate(Blades, sample);
            templates.Add(t);
        }

        /**
         * Learn a rejection threshold for each template.
         * Call after all templates have been added.
         *
         * See explanation in jackknife_train to understand
         * the input parameters.
         */
        public void Train(int gpsrN, int gpsrR, double beta)
        {
            int template_cnt = templates.Count;
            List<Distributions> distributions = new List<Distributions>();
            List<Vector> synthetic = new List<Vector>();

            double worst_score = 0.0;

            Random rand = new Random();

            //
            // Create negative samples.
            //
            for (int i = 0; i < 1000; i++)
            {
                synthetic.Clear();

                // Splice two samples together
                // to create one negative sample.
                for (int j = 0; j < 2; j++)
                {
                    int t = rand.Next(template_cnt);
                    Sample s = templates[t].Sample;

                    int len = s.Trajectory.Count;
                    int start = rand.Next(len / 2);

                    for (int kk = 0;kk < len / 2; kk++)
                    {
                        synthetic.Add(s.Trajectory[start + kk]);
                    }
                }

                JackknifeFeatures features = new JackknifeFeatures(Blades, synthetic);

                // and score it
                for (int t = 0; t < template_cnt; t++)
                {
                    double score = DTW(
                        features.Vecs,
                        templates[t].Features.Vecs);

                    if (worst_score < score)
                        worst_score = score;

                    if (i > 50)
                        distributions[t].AddNegativeScore(score);
                }

                // Generate a few samples to get an estimate
                // of worst possible score.
                if (i != 50)
                    continue;

                // allocate distributions
                for (int t = 0; t < template_cnt; t++)
                {
                    distributions.Add(new Distributions(worst_score, 1000));
                }
            }

            //
            // Create positive examples.
            //
            for (int t = 0; t < template_cnt; t++)
            {
                for (int i = 0; i < 1000; i++)
                {
                    synthetic.Clear();

                    // Create a synthetic variation of the sample.
                    synthetic = Mathematics.GPSR(templates[t].Sample.Trajectory, gpsrN, 0.25, gpsrR);

                    JackknifeFeatures features = new JackknifeFeatures(Blades, synthetic);

                    // and score it
                    double score = DTW(features.Vecs, templates[t].Features.Vecs);

                    distributions[t].AddPositiveScore(score);
                }
            }

            //
            // Now extract the rejection thresholds.
            //
            for (int t = 0; t < template_cnt; t++)
            {
                double threshold = distributions[t].RejectionThreshold(beta);
                JackknifeTemplate temp = templates[t];
                temp.RejectionThreshold = threshold;
                templates[t] = temp;
            }
        }

        /**
        * Determine gesture class of the given sample.
        * The gesture id is returned.
        *
        * This function can be modified to return an n-best list, but
        * because of early rejection, such a list may not make sense.
        *
        */
        public int Classify(List<Vector> trajectory)
        {
            JackknifeFeatures features = new JackknifeFeatures(Blades, trajectory);
            int templateCnt = templates.Count;
            for (int t = 0; t < templateCnt; t++)
            {
                double cf = 1;
                if (Blades.CFAbsDistance)
                {
                    cf *= 1 / Math.Max(0.01, features.Abs.Dot(templates[t].Features.Abs));
                }

                if (Blades.CFBbWidths)
                {
                    cf *= 1 / Math.Max(0.01, features.Bb.Dot(templates[t].Features.Bb));
                }

                JackknifeTemplate temp = templates[t];
                temp.CF = cf;
                templates[t] = temp;

                if (Blades.LowerBound)
                {
                    JackknifeTemplate tempLB = templates[t];
                    tempLB.LB = cf * LowerBound(features.Vecs, templates[t]);
                    templates[t] = tempLB;
                }
            }

            templates.Sort(new SortTemplate());

            double best = double.PositiveInfinity;
            int ret = -1;
            for (int t = 0; t < templateCnt; t++)
            {
                if (templates[t].LB > templates[t].RejectionThreshold)
                    continue;

                if (templates[t].LB > best)
                    continue;

                double score = templates[t].CF;
                score *= DTW(features.Vecs, templates[t].Features.Vecs);
                if (score > templates[t].RejectionThreshold)
                    continue;
                if (score < best)
                {
                    best = score;
                    ret = templates[t].GestureId;
                }
            }

            return ret;

        }

        public int Classify(Sample sample)
        {
            return Classify(sample.Trajectory);
        }

        /**
         * Find the lower bound score for a candidate
         * against a given template.
         */
        private double LowerBound(List<Vector> vecs, JackknifeTemplate t)
        {
            double lb = 0.0; // lower bound
            int component_cnt = vecs[0].Size;

            for (int i = 0; i < vecs.Count; i++)
            {
                double cost = 0.0;

                for (int j = 0; j < component_cnt; j++)
                {
                    if (Blades.InnerProduct)
                    {
                        if (vecs[i][j] < 0.0)
                        {
                            cost += vecs[i][j] * t.Lower[i][j];
                        }
                        else
                        {
                            cost += vecs[i][j] * t.Upper[i][j];
                        }
                    }
                    else if (Blades.EuclideanDistance)
                    {
                        double diff = 0.0;

                        if (vecs[i][j] < t.Lower[i][j])
                        {
                            diff = vecs[i][j] - t.Lower[i][j];
                        }
                        else if (vecs[i][j] > t.Upper[i][j])
                        {
                            diff = vecs[i][j] - t.Upper[i][j];
                        }

                        cost += (diff * diff);
                    }
                    else
                        throw new Exception("Should not get here!");
                }

                // inner products are bounded
                if (Blades.InnerProduct)
                    cost = 1.0 - Math.Min(1.0, Math.Max(-1.0, cost));

                lb += cost;
            }

            return lb;
        }

        /**
         * The DTW algorithm. Awesome!
         */
        private double DTW(List<Vector> v1, List<Vector> v2)
        {
            // Initialize the cost matrix if not already initialize
            if (cost == null)
            {
                cost = new List<List<double>>();

                for (int i = 0; i < v1.Count + 1; i++)
                {
                    List<double> row = new List<double>();

                    for (int j = 0; j < v2.Count + 1; j++)
                        row.Add(double.PositiveInfinity);

                    cost.Add(row);
                }
            }
            else
            {
                for (int i = 0; i < v1.Count + 1; i++)
                    for (int j = 0; j < v2.Count + 1; j++)
                        cost[i][j] = double.PositiveInfinity;
            }


            cost[0][0] = 0;

            // using DP to find solution
            for (int i = 1; i <= v1.Count; i++)
            {
                for (int j = Math.Max(1, i - Blades.Radius); j <= Math.Min(v2.Count, i + Blades.Radius); j++)
                {
                    // pick minimum cost path (neighbor) to
                    // extend to this ii, jj element
                    double minimum = Math.Min(Math.Min(cost[i - 1][j],      // repeat v1 element
                                                       cost[i][j - 1]),     // repeat v2 element
                                                       cost[i - 1][j - 1]); // don't repeat either
                    cost[i][j] = minimum;
                    if (Blades.InnerProduct)
                    {
                        cost[i][j] += 1 - v1[i - 1].Dot(v2[j - 1]);
                    }
                    else if (Blades.EuclideanDistance)
                    {
                        cost[i][j] += v1[i - 1].L2Norm2(v2[j - 1]);
                    }
                    else
                        throw new Exception("Should not get here!");
                }
            }

            return cost[v1.Count][v2.Count];
        }
    }
}
